csharp/src/Ice/Proxy.cs

```diff
@@ -415,21 +415,36 @@ public static class Proxy
                                                                progressWrapper,
                                                                cancel).ConfigureAwait(false);
 
-                            switch (response.ReplyStatus)
+                            // TODO: the observer RemoteException notification is not quite correct, since with the 1.1
+                            // encoding it suppresses (on purpose) the system exceptions.
+                            if ((response.Encoding == Encoding.V1_1 &&
+                                    response.ReplyStatus == ReplyStatus.UserException) ||
+                                (response.Encoding != Encoding.V1_1 && response.ResultType == ResultType.Failure))
                             {
-                                case ReplyStatus.OK:
-                                    break;
-                                case ReplyStatus.UserException:
-                                    observer?.RemoteException();
-                                    break;
-                                case ReplyStatus.ObjectNotExistException:
-                                case ReplyStatus.FacetNotExistException:
-                                case ReplyStatus.OperationNotExistException:
-                                    throw response.ReadDispatchException();
-                                case ReplyStatus.UnknownException:
-                                case ReplyStatus.UnknownLocalException:
-                                case ReplyStatus.UnknownUserException:
-                                    throw response.ReadUnhandledException();
+                                observer?.RemoteException();
+                            }
+
+                            // TODO: revisit
+                            // We throw here the 1.1 system exceptions, as they are used for retries
+                            if (response.Encoding == Encoding.V1_1 &&
+                                (byte)response.ReplyStatus > (byte)ReplyStatus.UserException)
+                            {
+                                InputStream istr;
```

Comment:
Why can't we call response.ReadException?

Reply:
I added a new internal `response.ThrowIfSystemException`. There is some duplication with ReadException, but we need to revisit this code in any case.